libcaliptra/src/caliptra_api.c

```diff
@@ -0,0 +1,277 @@
+    caliptra_fuse_array_write(GENERIC_AND_FUSE_REG_FUSE_OWNER_PK_HASH_0, fuses->owner_pk_hash, sizeof(fuses->owner_pk_hash));
+    caliptra_fuse_write(GENERIC_AND_FUSE_REG_FUSE_FMC_KEY_MANIFEST_SVN, fuses->fmc_key_manifest_svn);
+    caliptra_fuse_array_write(GENERIC_AND_FUSE_REG_FUSE_FMC_KEY_MANIFEST_SVN, fuses->runtime_svn, sizeof(fuses->runtime_svn));
+    caliptra_fuse_write(GENERIC_AND_FUSE_REG_FUSE_ANTI_ROLLBACK_DISABLE, (uint32_t)fuses->anti_rollback_disable);
+    caliptra_fuse_array_write(GENERIC_AND_FUSE_REG_FUSE_IDEVID_CERT_ATTR_0, fuses->idevid_cert_attr, sizeof(fuses->idevid_cert_attr));
+    caliptra_fuse_array_write(GENERIC_AND_FUSE_REG_FUSE_IDEVID_MANUF_HSM_ID_0, fuses->idevid_manuf_hsm_id, sizeof(fuses->idevid_manuf_hsm_id));
+    caliptra_fuse_write(GENERIC_AND_FUSE_REG_FUSE_LIFE_CYCLE, (uint32_t)fuses->life_cycle);
+
+    // Write to Caliptra Fuse Done
+    caliptra_write_u32(CALIPTRA_TOP_REG_GENERIC_AND_FUSE_REG_CPTRA_FUSE_WR_DONE, 1);
+
+    // No longer ready for fuses
+    if (caliptra_ready_for_fuses())
+        return -EIO;
+
+    return 0;
+}
+
+/**
+ * caliptra_bootfsm_go
+ *
+ * Initiate caliptra hw startup
+ *
+ * @return 0 if successful
+ */
+int caliptra_bootfsm_go()
+{
+    // Write BOOTFSM_GO Register
+    caliptra_write_u32(CALIPTRA_TOP_REG_GENERIC_AND_FUSE_REG_CPTRA_BOOTFSM_GO, 1);
+
+    //TODO: Check registers/provide async completion mechanism
+
+    return 0;
+}
+
+/**
+ * caliptra_mailbox_write_fifo
+ *
+ * Transfer contents of buffer into the mailbox FIFO
+ *
+ * @param[in] buffer Pointer to a valid caliptra_buffer struct
+ *
+ * @return int -EINVAL if the buffer is too large.
+ */
+static int caliptra_mailbox_write_fifo(struct caliptra_buffer *buffer)
+{
+    // Check against max size
+    const uint32_t MBOX_SIZE = (128u * 1024u);
+
+    if (buffer->len > MBOX_SIZE) {
+        return -EINVAL;
+    }
+
+    // Write DLEN
+    caliptra_mbox_write_dlen(buffer->len);
+
+    uint32_t remaining_len = buffer->len;
+    uint32_t *data_dw = (uint32_t *)buffer->data;
+
+    // Copy DWord multiples
+    while (remaining_len > sizeof(uint32_t)) {
+        caliptra_mbox_write(MBOX_CSR_MBOX_DATAIN, *data_dw++);
+        remaining_len -= sizeof(uint32_t);
+    }
+
+    // if un-aligned dword reminder...
+    if (remaining_len) {
+        uint32_t data = 0;
+        memcpy(&data, data_dw, remaining_len);
+        caliptra_mbox_write(MBOX_CSR_MBOX_DATAIN, data);
+    }
+
+    return 0;
+}
+
+/**
+ * caliptra_mailbox_read_buffer
+ *
+ * Read a mailbxo FIFO into a buffer
+ *
+ * @param[in] buffer A pointer to a valid caliptra_buffer struct
+ *
+ * @return int 0 if successful, -EINVAL if the buffer is too small or the buffer pointer is invalid.
+ */
+static int caliptra_mailbox_read_buffer(struct caliptra_buffer *buffer)
+{
+    uint32_t remaining_len = caliptra_mbox_read_dlen();
+
+    // Check we have enough room in the buffer
+    if (buffer->len < remaining_len || !buffer->data)
```

Comment:
Probably a good idea to also check if the buffer ptr is valid.

Reply:
Can do this too